### DISCUSSION TITLE
Guild.Members doesn't give me all the members

### DISCUSSION THREAD

**Discussion by DjangoLowe** [DSharpPlus/DSharpPlus#879] [Support] [answered]
Hi there,

I'm trying to write a command that measures the activity of the members in a role and there doesn't seem to be a way of getting the members of a role.  When I get the all I get is myself and the bot in the member's member.  any way I can get all the members of a role?

**Comment by CloudTheWolf** [comment]
You could do something like this to generate a list of all members with role x

```
            var members = ctx.Guild.Members.Values.ToArray<DiscordMember>();
            var membersWithRoleX = new List<KeyValuePair<ulong, string>>();
            var role = ctx.Guild.GetRole({{PUT_ROLE_ID_HERE}});
            foreach (var member in members.Where(member => member.Roles.Contains(role)))
            {
                Console.WriteLine($"{member.Username} has role");
                membersWithRoleX.Add(new KeyValuePair<ulong, string>(member.Id, member.Username));
                //Do Something With This Info Here...
            }
            // Do something with the list, Eg DM it.
```

**Comment by CabbageAdi** [comment, marked as answer]
`Guild.Members` usually gives an incomplete list because of caching. Use `GetAllMembersAsync()` to get a full list.

**Comment by c0nnex** [comment]
Sorry to warm this old post up.
GetAllMembersAsync() does not return all members sometimes. Just had the case that about 15 long term members where missing from the list.
DSharpPlus, version 4.3.0-nightly-01175, net5.0
